### PR TITLE
sql: Add unit tests for planning inside the new schema changer

### PR DESCRIPTION
**Pull request by fqazi**
Previously, the new schema changer did not have any unit
tests covering the planning capability. This was inadequate,
because we had no way of detect if plans were regressing with
changes or enhancements. To address this, this patch adds
basic tests to see if the operators/dependencies for a given
command are sane.

Release note: None

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiW-KjBzpMfuVBPzGo:-MaiW-KjBzpMfuVBPzGp:bav63cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f704026181802c51473baa4a131375217a50010/pkg/sql/schemachanger/scplan/plan_test.go#L138)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 		data, err := json.Marshal(node)
> 		require.NoError(t, err)
> 		depMap := make(map[string]interface{})
> 		err = gojson.Unmarshal(data, &depMap)
> 		require.NoError(t, err)
> ```
</details>

Let's replace this all with `scgraphviz.toMap` 
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiVd_1EbQL_OePAlSw:-MaiVd_1EbQL_OePAlSx:b-359v0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f704026181802c51473baa4a131375217a50010/pkg/sql/schemachanger/scplan/testdata/alter_table#L18)):*
> ```
>   name: j
>   nullable: true
>   pg_attribute_num: 0
> ```

Getting these default values is painful. Maybe we should just export `scgraphviz.toMap` I think that thing does some pretty good stuff pruning the protobuf zero fields and objects.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiW-KjBzpMfuVBPzGo:-MajJ9BJ76InXcC279Bv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f704026181802c51473baa4a131375217a50010/pkg/sql/schemachanger/scplan/plan_test.go#L138)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		data, err := json.Marshal(node)
> 		require.NoError(t, err)
> 		depMap := make(map[string]interface{})
> 		err = gojson.Unmarshal(data, &depMap)
> 		require.NoError(t, err)
> ```
Let's replace this all with `scgraphviz.toMap` 
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiVd_1EbQL_OePAlSw:-MajJ9iv3h3IK99CvUDz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f704026181802c51473baa4a131375217a50010/pkg/sql/schemachanger/scplan/testdata/alter_table#L18)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Getting these default values is painful. Maybe we should just export `scgraphviz.toMap` I think that thing does some pretty good stuff pruning the protobuf zero fields and objects. 
</blockquote></details>

Done.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MajRKCH91onv-LfftG-:-MajRKCH91onv-LfftG0:b-s0xccm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a0e5ea1df82d6a834fa59fbd53913b2184228e2/pkg/sql/schemachanger/scplan/testdata/alter_table#L8)):*
> ```
> ALTER TABLE defaultdb.foo ADD COLUMN j INT
> ----
> Stage 0
> ```

What if we did something like below that the eye can see and visually understanding grouping and hierarchy better? It'll be a bit more complex but I don't think it'll be a ton more complex. I'm far from wedded from the below format but I'm trying to make something that looks somewhat like the textual explain. My sense is it will be valuable to have a textual output that we get comfortable with. Anything we do here we can reuse later. 


```
 ├─Stage 0
 │   ├─*scop.MakeAddedColumnDeleteOnly
 │   │    Column:
 │   │        ID: 2
 │   │        Name: j
 │   │        Nullable: true
 │   │        Type:
 │   │            family: IntFamily
 │   │            oid: 20
 │   │            width: 64
 │   │        FamilyName: primary
 │   │    TableID: 52
 │   └─*scop.MakeAddedIndexDeleteOnly
 │        Index:
 │            EncodingType: 1
 │            ID: 2
 │            KeyColumnDirections:
 │            - 0
 │            KeyColumnIDs:
 │            - 1
 │            KeyColumnNames:
 │            - i
 │            Name: new_primary_key
 │            StoreColumnIDs:
 │            - 2
 │            StoreColumnNames:
 │            - j
 │            Unique: true
 │            Version: 3
 │        TableID: 52
 ├─Stage 1
 │   ├─*scop.MakeAddedColumnDeleteAndWriteOnly
```

File: pkg/sql/schemachanger/scplan/testdata/alter_table [hunk 1]
```diff
@@ -0,0 +1,559 @@
+create-table
+CREATE TABLE defaultdb.foo (i INT PRIMARY KEY)
+----
+
+ops
+ALTER TABLE defaultdb.foo ADD COLUMN j INT
+----
+Stage 0
+	*scop.MakeAddedColumnDeleteOnly
```
Comment (ajwerner) [inline pkg/sql/schemachanger/scplan/testdata/alter_table:9]: I'm getting nitpicky now: it'd be nice if the yaml indentation levels matched the outer indentation levels.

File: pkg/sql/schemachanger/scplan/testdata/drop_sequence [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	*scop.CreateGcJobForDescriptor
+		DescID: 52
+	*scop.DrainDescriptorName
+		TableID: 52
+
+
+deps
+DROP SEQUENCE defaultdb.SQ1 CASCADE
+----
+Dependency From:
```
Comment (ajwerner) [inline pkg/sql/schemachanger/scplan/testdata/drop_sequence:55]: It'd be nice if this got a similar indentation treatment.

**Review by ajwerner — COMMENTED** [review]
Much better.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scplan/testdata/alter_table, line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiIpCbC5jrCmCjO6c5-r3-9:-ManCp81C9HvILlieTzc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1163b3281262b24c7ef60b98d25cea8bf4262633/pkg/sql/schemachanger/scplan/testdata/alter_table#L9)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm getting nitpicky now: it'd be nice if the yaml indentation levels matched the outer indentation levels.
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/testdata/drop_sequence, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiIpCbC5jrCmCjO6c6-r3-55:-ManCphlEUMMca5Nr8wB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1163b3281262b24c7ef60b98d25cea8bf4262633/pkg/sql/schemachanger/scplan/testdata/drop_sequence#L55)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It'd be nice if this got a similar indentation treatment. 
</blockquote></details>

Done.

**Review by ajwerner — COMMENTED** [review]
nit: lowercase `Add` in the commit message.

> Reviewed 1 of 5 files at r1, 1 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-:-MankMzk-YtOrGGJxVvf:b9plve8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/schemachanger/scplan/testdata/drop_sequence, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MaiIpCbC5jrCmCjO6c6-r3-55:-ManklNW01lhNY97luAU:b-hnzn82) ([raw file](https://github.com/cockroachdb/cockroach/blob/1163b3281262b24c7ef60b98d25cea8bf4262633/pkg/sql/schemachanger/scplan/testdata/drop_sequence#L55)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think we still want more iteration here. It's hard to see the graph. What if we collect the set of nodes involved in edges, print them much like we print them in the stages thing, and then we reference them with some shorthand. 

I think this is a place where the attributes stuff you're adding elsewhere will really come in handy. What if we did it as the set of attributes in a compact representation and then used arrows.

Like imagine the top level was an object with two keys, nodes and edges. Nodes was a map of the "fingerprint" serialization we come up with based on the attributes for the element to the element and then edges is either a map of fingerprint to fingerprint or a list of strings with an arrow between them.
defaultExpression(tableID:52, columnID: 2):ABSENT -> table(tableID: 52):DELETE_ONLY`
___
*[pkg/sql/schemachanger/scplan/testdata/drop_sequence, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-ManldF63GXPklL012-v:-ManldF63GXPklL012-w:b-eie3ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/38531fad1318699d534950bf9f1c861d9cd8bd0a/pkg/sql/schemachanger/scplan/testdata/drop_sequence#L56)):*
> ```
> ----
> Dependency From:
>   State: 1
> ```

it'd be nice to have strings for the states

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scplan/testdata/drop_sequence, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-ManldF63GXPklL012-v:-MbC4wV_9tWAAPlbC9hY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/38531fad1318699d534950bf9f1c861d9cd8bd0a/pkg/sql/schemachanger/scplan/testdata/drop_sequence#L56)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it'd be nice to have strings for the states
</blockquote></details>

Done.

**Review by ajwerner — COMMENTED** [review]
> Reviewed 1 of 3 files at r4, 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 87 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-Mb3rc6L0BcsI32YOnkv:-Mb3rc6L0BcsI32YOnkw:b3ahw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/38531fad1318699d534950bf9f1c861d9cd8bd0a/pkg/sql/schemachanger/scplan/plan_test.go#L87)):*
> ```
> 
> 				fallthrough
> 			case "ops":
> ```

Could be another `,` instead of `fallthrough`
___
*[pkg/sql/schemachanger/scplan/testdata/drop_sequence, line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbH1vfG9Znop7BP_D9j:-MbH1vfG9Znop7BP_D9k:b-8tri21) ([raw file](https://github.com/cockroachdb/cockroach/blob/a150367bf330916e9826e4b988d9dac0a072e84b/pkg/sql/schemachanger/scplan/testdata/drop_sequence#L55)):*
> ```
> DROP SEQUENCE defaultdb.SQ1 CASCADE
> ----
> *scpb.DefaultExpression[ Type: 8, DescID: 53, ColumnID: 2 ]:ABSENT->*scpb.Sequence[ Type: 7, DescID: 52 ]:DELETE_ONLY
> ```

I think it would also be nice to get the type here. I'd go one further and say that I don't think we need to print the type given it's encoded in `8`.

I could also see with this a new format for the `Attributes`. I'm far from wedded to the `[ ... ]`. Maybe `<type>: {<key>: <value> }` so `DefaltExpression: {DescID: 53, ColumnID: 2}`? That ends up being yaml.

**Review by ajwerner — COMMENTED** [review]
just a couple more minor formatting things and it'll be good to go. 

> Reviewed 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-:-MbHZ7Ju-8CzUGT56oxO:b-4bucbc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/sql/schemachanger/scpb/attribute.go, line 162 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbHWARV5OUjn-eAykXx:-MbHWARV5OUjn-eAykXy:b-xwz1xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/159668a56e29354c953866d5d25ab6899bf3f254/pkg/sql/schemachanger/scpb/attribute.go#L162)):*
> ```Go
> 		}
> 	}
> 	result.WriteString(" }")
> ```

Can we lose this trailing ` `?
___
*[pkg/sql/schemachanger/scpb/attribute_test.go, line 41 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbHWPliEgLKRSdARANd:-MbHWPliEgLKRSdARANe:bbcwvsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/159668a56e29354c953866d5d25ab6899bf3f254/pkg/sql/schemachanger/scpb/attribute_test.go#L41)):*
> ```Go
> 	// work.
> 	typeBackRef := TypeReference{TypeID: 3, DescID: 1}
> 	expectedStr = `*scpb.TypeReference:{DescID: 3, DepID: 1 }`
> ```

nit: how do you feel about trimming `*scpb` in `ElementIDToString`? I think it'll make some things cleaner.
___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 154 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbHWf8l54yEwfWs6fGB:-MbHWf8l54yEwfWs6fGC:b1ka4l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/159668a56e29354c953866d5d25ab6899bf3f254/pkg/sql/schemachanger/scplan/plan_test.go#L154)):*
> ```
> 			stages += fmt.Sprintf("%s:%s->%s:%s\n",
> 				toAttr, de.To().State,
> 				fromAttr, de.From().State)
> ```
What if this also were made to be valid yaml (also moving to strings.Builder).

```
	var stages string.Builder
	plan.Graph.ForEachNode(func(n *scpb.Node) error {
		return plan.Graph.ForEachDepEdgeFrom(n, func(de *scgraph.DepEdge) error {
			toAttr := de.To().Element().GetAttributes()
			fromAttr := de.From().Element().GetAttributes()
                        fmt.Fprintf(&stages, "- from: [%s, %s]\n", toAttr, de.To().State),
                        fmt.Fprintf(&stages, "  to:   [%s, %s]\n", fromAttr, de.From().State)
                        return nil
```
Such that what is now:

```
*scpb.View:{DescID: 54 }:DELETE_ONLY->*scpb.View:{DescID: 53 }:DELETE_ONLY
*scpb.TypeReference:{DescID: 57, DepID: 59 }:ABSENT->*scpb.View:{DescID: 59 }:DELETE_ONLY
```

will be:
```yaml
- from: [View:{DescID: 54}, DELETE_ONLY]
  to:   [View:{DescID: 53}, DELETE_ONLY]
- from: [TypeReference:{DescID: 57, DepID: 59}, ABSENT]
  to:   [View:{DescID: 59}, DELETE_ONLY]
```

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/scpb/attribute.go, line 162 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbHWARV5OUjn-eAykXx:-MbI3ekg4U2oqgEg09z6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159668a56e29354c953866d5d25ab6899bf3f254/pkg/sql/schemachanger/scpb/attribute.go#L162)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can we lose this trailing ` `?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/scplan/plan_test.go, line 154 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbHWf8l54yEwfWs6fGB:-MbI3jju6MuUK33KtyiV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/159668a56e29354c953866d5d25ab6899bf3f254/pkg/sql/schemachanger/scplan/plan_test.go#L154)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			stages += fmt.Sprintf("%s:%s->%s:%s\n",
> 				toAttr, de.To().State,
> 				fromAttr, de.From().State)
> ```
What if this also were made to be valid yaml (also moving to strings.Builder).

```
	var stages string.Builder
	plan.Graph.ForEachNode(func(n *scpb.Node) error {
		return plan.Graph.ForEachDepEdgeFrom(n, func(de *scgraph.DepEdge) error {
			toAttr := de.To().Element().GetAttributes()
			fromAttr := de.From().Element().GetAttributes()
                        fmt.Fprintf(&stages, "- from: [%s, %s]\n", toAttr, de.To().State),
                        fmt.Fprintf(&stages, "  to:   [%s, %s]\n", fromAttr, de.From().State)
                        return nil
```
Such that what is now:

```
*scpb.View:{DescID: 54 }:DELETE_ONLY->*scpb.View:{DescID: 53 }:DELETE_ONLY
*scpb.TypeReference:{DescID: 57, DepID: 59 }:ABSENT->*scpb.View:{DescID: 59 }:DELETE_ONLY
```

will be:
```yaml
- from: [View:{DescID: 54}, DELETE_ONLY]
  to:   [View:{DescID: 53}, DELETE_ONLY]
- from: [TypeReference:{DescID: 57, DepID: 59}, ABSENT]
  to:   [View:{DescID: 59}, DELETE_ONLY]
```
</blockquote></details>

Done.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 6 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-:-MbID7H2BscQSGOT4yX2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[pkg/sql/schemachanger/scpb/elements.go, line 80 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/65780#-MbICBqw42grLkqIetuO:-MbICBqw42grLkqIetuP:bx4aksa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e9a34a64a2f28f2e93a7eec1de377bd209f3968/pkg/sql/schemachanger/scpb/elements.go#L80)):*
> ```Go
> 		}
> 		typeToElementID[f.Type] = ElementTypeID(id)
> 		elementIDToString[ElementTypeID(id)] = strings.ReplaceAll(f.Type.String(), "*scpb.", "")
> ```

nit: `strings.TrimPrefix`

**Comment by fqazi** [comment]
TFTR @ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by rickystewart** [comment]
(drive-by) you need to `make bazel-generate`, this will not pass CI

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by ajwerner** [comment]
darn cla bot seems sad. https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73203936/CLA+Assistant+glitch+in+PRs or push a no-op commit

**Comment by fqazi** [comment]
bors r=ajwerner

**Comment by fqazi** [comment]
> darn cla bot seems sad. https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73203936/CLA+Assistant+glitch+in+PRs or push a no-op commit

Thanks @ajwerner

**Pull request merged**